spec/sneakers/integration_spec.rb

```diff
@@ -100,8 +85,352 @@ def start_worker(w)
+    end
+    let(:redis) { Redis.new }
+    let(:queue_name) { "integration_#{rand(10**36).to_s(36)}" }
+    let(:exchange_name) { "integration_#{rand(10**36).to_s(36)}" }
+
+    before :each do
+      skip unless ENV['INTEGRATION']
+
+      begin
+        rabbitmq_client.overview
+      rescue
+        puts 'Rabbitmq admin seems to not exist? You better be running this on'\
+          "Travis or Docker. proceeding.\n#{$ERROR_INFO}"
+        skip
+      end
+
+      begin
+        redis.info
+      rescue
+        puts 'Redis seems to not exist? You better be running this on'\
+          'Travis or Docker.'
+        skip
+      end
+
+      cleanup_rabbitmq(rabbitmq_client)
+      cleanup_redis(redis)
+      prepare_sneakers(exchange: exchange_name)
+
+      worker.from_queue(queue_name, worker_opts)
+      @worker_pid = start_worker(worker)
+    end
+
+    after :each do
+      Process.kill('TERM', @worker_pid)
+
+      cleanup_rabbitmq(rabbitmq_client)
+      cleanup_redis(redis)
+    end
+
+    describe 'with defaults for maxretry handler' do
+      let(:worker) { MaxretryWorker }
+      let(:worker_opts) do
+        {
+          handler:         Sneakers::Handlers::Maxretry,
+          retry_max_times: 2,
+          retry_timeout:   100,
+          arguments: {
+            'x-dead-letter-exchange' => "#{queue_name}-retry"
+          }
+        }
+      end
+
+      it 'creates the required exchanges' do
+        expected_exchange_names = [
+          exchange_name,
+          "#{queue_name}-error",
+          "#{queue_name}-retry",
+          "#{queue_name}-retry-requeue"
+        ]
+        exchange_names = rabbitmq_client.list_exchanges.map(&:name)
+
+        expected_exchange_names.each do |expected_exchange_name|
+          assert_includes(exchange_names, expected_exchange_name)
+        end
+      end
+
+      it 'creates the required queues' do
+        expected_queue_names = [
+          queue_name,
+          "#{queue_name}-error",
+          "#{queue_name}-retry"
+        ]
+        queue_names = rabbitmq_client.list_queues.map(&:name)
+
+        expected_queue_names.each do |expected_queue_name|
+          assert_includes(queue_names, expected_queue_name)
+        end
+      end
+
+      describe 'when worker allways fails' do
+        let(:worker) { FailingMaxretryWorker }
+
+        before do
+          Sneakers::Publisher.new(
+            exchange: exchange_name
+          ).publish(
+            'foo',
+            routing_key: queue_name
+          )
+        end
+
+        it 'has a message in the error queue' do
+          # wait for failing message
+          sleep REQUEUING_WAIT_TIME
+
+          message = get_message_from_queue("#{queue_name}-error")
+
+          refute_nil(message.first)
+        end
+
+        it 'has been retried twice' do
+          # wait for failing message
+          sleep REQUEUING_WAIT_TIME
+
+          message = get_message_from_queue("#{queue_name}-error")
+
+          # Hmmmmm... should be 2
```

Comment:
The 3 is correct here. It is retried twice, which means a worker sees it 3 times total, yielding the 3 attempts.